### ISSUE TITLE
[feature request] allow adapters in more locations

### ISSUE THREAD

**Issue by ygrek** [mjambon/atd#130]
Usecase:
Some api returns a field in record sometimes as string and sometimes as string list. This can be currently hacked around with `json wrap` but looks like adapter could provide a natural solution when applied on record field position.

**Comment by gtrak** [comment]
I could use this on plain lists, was confused that it didn't work given the documentation

**Comment by mjambon** [comment]
@gtrak I agree with the need to support the feature on lists. I actually thought we did so it's possible that I wrote that somewhere (in an announcement maybe?). The reference manual is correct, though (https://atd.readthedocs.io/en/latest/atdgen.html#field-adapter-ocaml).

**Comment by gtrak** [comment]
I wasn't able to make it work, then traced it to the source and the emitter seems specific to record and variant types?  If we don't yet support it, it might be better to syntax error than silently fail?  I might take a look at this eventually since my workaround was really really ugly.

**Comment by mjambon** [comment]
@gtrak the technical issue is to avoid code duplication. Right now the `<json adapter.ocaml=...>` annotation is translated separately for each kind of node of the AST. In `oj_emit.ml`, we have:
```ocaml
      let adapter = j.json_sum_adapter in
      write_with_adapter adapter standard_writer
```
and for records we have:
```ocaml
      let adapter = j.json_record_adapter in
      write_with_adapter adapter standard_writer
```

Having `wrap` nodes in the core atd syntax is one way we solved this problem in the past, but I don't think they're suitable for use with json adapters. I also hope we don't have to create yet another kind of artificial node in the ast.
